### PR TITLE
Skip all hidden directories in ``CatalogRepository``

### LINKED ISSUE #11666: Not skipping all dot directories
### Describe the bug

In [this loop](https://github.com/sphinx-doc/sphinx/blob/fcc38997f1d9b728bb4ffc64fc362c7763a4ee25/sphinx/util/i18n.py#L104), if there are at least 2 directories that start with `.` and are adjacent in the `dirnames` list, then one of the is not skipped. This happens because the `dirnames.remove(dirname)` changes the list it is iterating on.

### How to Reproduce

Just having two directories like `.dotdir1` and `.dotdir2` should trigger this problem.

### Environment Information

```text
Platform:              linux; (Linux-5.19.0-46-generic-x86_64-with-glibc2.35)
Python version:        3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0])
Python implementation: CPython
Sphinx version:        7.2.5
Docutils version:      0.20.1
Jinja2 version:        3.1.2
Pygments version:      2.16.1
```

### Sphinx extensions

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AryazE**
Subject: Fixes #11666 

### Feature or Bugfix

- Bugfix

### Purpose
This PR fixes the problem with not skipping all dot directories in `CatalogRepository.pofiles`. 

### Detail
The fix replace the removal of `dirname` in the loop with collecting the names of directories to remove and then removing them without iterating over `dirnames`.

### Relates
- #11666

### CODE REVIEW

**Review by picnixz — APPROVED** [review]
Good catch!

File: sphinx/util/i18n.py [hunk 1]
```diff
@@ -99,9 +99,9 @@ def pofiles(self) -> Generator[tuple[str, str], None, None]:
             basedir = path.join(locale_dir, self.language, 'LC_MESSAGES')
             for root, dirnames, filenames in os.walk(basedir):
                 # skip dot-directories
-                for dirname in dirnames:
-                    if dirname.startswith('.'):
-                        dirnames.remove(dirname)
+                dotdirs = [d for d in dirnames if d.startswith('.')]
+                for dotdir in dotdirs:
+                    dirnames.remove(dotdir)
```
Comment (AA-Turner) [inline sphinx/util/i18n.py]:
```suggestion
                for dirname in [d for d in dirnames if d.startswith('.')]:
                    dirnames.remove(dirname)
```

**Comment by AA-Turner** [comment]
Please add an entry to CHANGES

A

File: CHANGES.rst [hunk 2]
```diff
@@ -20,6 +20,8 @@ Bugs fixed
   Patch by Vinay Sajip.
 * #11622: Ensure that the order of keys in ``searchindex.js`` is deterministic.
   Patch by Pietro Albini.
+* #11666: Fixed ignoring dot directories in ``CatalogRepository.pofiles``.
+  Patch by Aryaz Eghbali.
```
Comment (AA-Turner) [inline CHANGES.rst]:
```suggestion
* #11666: Skip all hidden directories in ``CatalogRepository.pofiles``.
  Patch by Aryaz Eghbali.
```

**Pull request merged**